torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,432 @@
+
+  /// Returns the number of chunks available in this reader.
+  virtual size_t get_chunk_count() = 0;
+
+  /// This will clear any internal state associate with this reader.
+  virtual void reset() = 0;
+};
+
+/// A class that contains a raw unwrapped batch unit. An unwrapped batch unit is
+/// the raw data without 'optional' wrapper. It can be a collection of images,
+/// utterances, e.t.c.
+template <typename UnwrappedBatch = std::vector<Example<>>>
+struct UnwrappedBatchData {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+
+  UnwrappedBatchData(UnwrappedBatchType data) : batch_data(std::move(data)) {}
+
+  UnwrappedBatchData(std::exception_ptr e) : exception(e) {}
+
+  /// batch data to return
+  UnwrappedBatchType batch_data;
+
+  /// exception pointer which captures any abnormal exceptions while creating the
+  /// batch.
+  std::exception_ptr exception;
+};
+
+/// BatchDataBuffer manages a queue of UnwrappedBatchData. After a new chunk is
+/// loaded, BatchDataBuffer splits it into small batches and push them into the
+/// queue. When get_batch is called from data loader, it pops cached batches and
+/// return. If the cache is empty, it either waits to load more chunks or return
+/// null if all chunks are loaded.
+template <
+    typename UnwrappedBatch = std::vector<Example<>>,
+    typename ExampleSampler = samplers::RandomSampler>
+class BatchDataBuffer {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+  using BatchType = torch::optional<UnwrappedBatchType>;
+  using BatchRequestType = typename ExampleSampler::BatchRequestType;
+
+  BatchDataBuffer(
+      size_t num_chunks,
+      size_t batch_size,
+      ExampleSampler example_sampler,
+      bool ignore_empty_chunk,
+      size_t cache_size)
+      : remaining_chunk_count_(num_chunks),
+        batch_size_(batch_size),
+        example_sampler_(std::move(example_sampler)),
+        ignore_empty_chunk_(ignore_empty_chunk),
+        queue_depth_(cache_size) {}
+
+  /// Return batch data from the queue. Called from the ChunkDataSet main
+  /// thread.
+  BatchType get_batch(size_t batch_size) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cvr_.wait(lock, [this] {
+      // wait till there is available data in the queue or if all chunks are
+      // loaded (i.e. the data set is exhausted for this epoch)
+      return (
+          this->total_example_count_in_queue_ >= batch_size_ ||
+          remaining_chunk_count_ == 0);
+    });
+    if (batch_queue_.empty()) {
+      lock.unlock();
+      AT_ASSERT(remaining_chunk_count_ == 0);
+
+      // All batches have been retrieved. Return an empty batch.
+      return nullopt;
+    }
+
+    auto batch_data = batch_queue_.front();
+    batch_queue_.pop();
+    if (batch_data.exception) {
+      throw WorkerException(batch_data.exception);
+    }
+
+    this->total_example_count_in_queue_ -= batch_data.batch_data.size();
+    lock.unlock();
+    cvw_.notify_all(); // notify all writers.
```

Comment:
Wouldn't it be better to `notify_one`? You've only taken a single item so there's no point in waking up 10 threads.

Reply:
We need to call notify_all to make it properly work. 

For example, if 10 threads are waiting in get_batch, and if we just notify 1 of them, the rest 9 threads will still be blocked by the waiting. At last, only 1 batch instead of 10 batches are returned. 

In a different case where the batch size is greater than chunk size, we need to notify all preloader threads to fill the queue, instead of just one.

In both cases (read and write), notify_all should be used. It may have a slight perf hit over notify_one, but this should be the right logic to use here.